common/persistence/sql/sqlplugin/mysql/visibility.go

```diff
@@ -29,252 +29,186 @@ import (
-
-	templateCreateWorkflowExecutionClosed = `INSERT INTO executions_visibility (` +
-		`namespace_id, workflow_id, run_id, start_time, execution_time, workflow_type_name, close_time, status, history_length, memo, encoding, task_queue) ` +
-		`VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) ` +
-		`ON DUPLICATE KEY UPDATE workflow_id = VALUES(workflow_id), start_time = VALUES(start_time), execution_time = VALUES(execution_time), workflow_type_name = VALUES(workflow_type_name), ` +
-		`close_time = VALUES(close_time), status = VALUES(status), history_length = VALUES(history_length), memo = VALUES(memo), encoding = VALUES(encoding), task_queue = VALUES(task_queue)`
-
-	// RunID condition is needed for correct pagination
-	templateConditions = ` AND namespace_id = ?
-		 AND start_time >= ?
-		 AND start_time <= ?
- 		 AND ((run_id > ? and start_time = ?) OR (start_time < ?))
-         ORDER BY start_time DESC, run_id
-		 LIMIT ?`
-
-	templateConditionsClosedWorkflows = ` AND namespace_id = ?
-	AND close_time >= ?
-	AND close_time <= ?
-	 AND ((run_id > ? and close_time = ?) OR (close_time < ?))
-	ORDER BY close_time DESC, run_id
-	LIMIT ?`
-
-	templateOpenFieldNames = `workflow_id, run_id, start_time, execution_time, workflow_type_name, status, memo, encoding, task_queue`
-	templateOpenSelect     = `SELECT ` + templateOpenFieldNames + ` FROM executions_visibility WHERE status = 1 `
-
-	templateClosedSelect = `SELECT ` + templateOpenFieldNames + `, close_time, history_length
-		 FROM executions_visibility WHERE status != 1 `
-
-	templateGetOpenWorkflowExecutions = templateOpenSelect + templateConditions
-
-	templateGetClosedWorkflowExecutions = templateClosedSelect + templateConditionsClosedWorkflows
-
-	templateGetOpenWorkflowExecutionsByType = templateOpenSelect + `AND workflow_type_name = ?` + templateConditions
-
-	templateGetClosedWorkflowExecutionsByType = templateClosedSelect + `AND workflow_type_name = ?` + templateConditionsClosedWorkflows
+var (
+	templateInsertWorkflowExecution = fmt.Sprintf(
+		`INSERT INTO executions_visibility (%s)
+		VALUES (%s)
+		ON DUPLICATE KEY UPDATE run_id = VALUES(run_id)`,
+		strings.Join(sqlplugin.DbFields, ", "),
+		sqlplugin.BuildNamedPlaceholder(sqlplugin.DbFields...),
+	)
 
-	templateGetOpenWorkflowExecutionsByID = templateOpenSelect + `AND workflow_id = ?` + templateConditions
+	templateInsertCustomSearchAttributes = `
+		INSERT INTO custom_search_attributes (
+			namespace_id, run_id, search_attributes
+		) VALUES (:namespace_id, :run_id, :search_attributes)
+		ON DUPLICATE KEY UPDATE run_id = VALUES(run_id)`
 
-	templateGetClosedWorkflowExecutionsByID = templateClosedSelect + `AND workflow_id = ?` + templateConditionsClosedWorkflows
+	templateUpsertWorkflowExecution = fmt.Sprintf(
+		`INSERT INTO executions_visibility (%s)
+		VALUES (%s)
+		%s`,
+		strings.Join(sqlplugin.DbFields, ", "),
+		sqlplugin.BuildNamedPlaceholder(sqlplugin.DbFields...),
+		buildOnDuplicateKeyUpdate(sqlplugin.DbFields...),
```

Comment:
Technically, insert is only used when workflow is created and upsert when there's an update. We use upsert instead of update because visibility is eventual consistency, and tasks might be processed out of order.

`BuildNamedPlaceholder: ["foo", "bar"] -> ":foo, :bar"`

`buildOnDuplicateKeyUpdate`: build the SQL statement to do the upsert.